modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotCheckProcess.java

```diff
@@ -0,0 +1,596 @@
+
+        if (log.isInfoEnabled())
+            log.info("Finished snapshot validation [req=" + ctx.req + ']');
+
+        GridFutureAdapter<SnapshotPartitionsVerifyTaskResult> clusterOpFut = clusterOpFuts.get(reqId);
+
+        if (clusterOpFut == null)
+            return new GridFinishedFuture<>();
+
+        assert results.values().stream().noneMatch(res -> res != null && res.metas != null);
+
+        if (ctx.req.allRestoreHandlers) {
+            try {
+                if (!F.isEmpty(errors))
+                    throw F.firstValue(errors);
+
+                Map<ClusterNode, Map<String, SnapshotHandlerResult<?>>> cstRes = mapCustomHandlersResults(results, ctx.req.nodes());
+
+                checkCustomHandlersResults(ctx.req.snapshotName(), cstRes);
+
+                clusterOpFut.onDone(new SnapshotPartitionsVerifyTaskResult(ctx.clusterMetas, null));
+            }
+            catch (Throwable err) {
+                clusterOpFut.onDone(err);
+            }
+        }
+        else {
+            Map<ClusterNode, Exception> errors0 = mapErrors(errors);
+
+            if (!F.isEmpty(results)) {
+                Map<ClusterNode, Map<PartitionKeyV2, PartitionHashRecordV2>> results0 = mapPartsHashes(results, ctx.req.nodes());
+
+                IdleVerifyResultV2 chkRes = SnapshotChecker.reduceHashesResults(results0, errors0);
+
+                clusterOpFut.onDone(new SnapshotPartitionsVerifyTaskResult(ctx.clusterMetas, chkRes));
+            }
+            else {
+                assert !errors0.isEmpty();
+
+                clusterOpFut.onDone(new IgniteSnapshotVerifyException(errors0));
+            }
+        }
+
+        return new GridFinishedFuture<>();
+    }
+
+    /** Phase 2 beginning.  */
+    private IgniteInternalFuture<SnapshotCheckResponse> validateParts(SnapshotCheckProcessRequest req) {
+        SnapshotCheckContext ctx;
+
+        // The context can be null, if a required node leaves before this phase.
+        if (!req.nodes().contains(kctx.localNodeId()) || (ctx = context(null, req.requestId())) == null || ctx.locMeta == null)
+            return new GridFinishedFuture<>();
+
+        File snpDir = kctx.cache().context().snapshotMgr().snapshotLocalDir(req.snapshotName(), req.snapshotPath());
+
+        ctx.fut = processPartitions(ctx.locMeta, snpDir, req.groups(), req.allRestoreHandlers, true);
+
+        return ctx.fut;
+    }
+
+    /** */
+    private GridFutureAdapter<SnapshotCheckResponse> processPartitions(
+        SnapshotMetadata meta,
+        File snpDir,
+        @Nullable Collection<String> groups,
+        boolean allRestoreHandlers,
+        boolean check
+    ) {
+        GridFutureAdapter<SnapshotCheckResponse> fut = new GridFutureAdapter<>();
+
+        CompletableFuture<? extends Map<?, ?>> workingFut;
+
+        IgniteSnapshotManager snpMgr = kctx.cache().context().snapshotMgr();
+
+        if (allRestoreHandlers) {
+            workingFut = CompletableFuture.supplyAsync(() -> {
```

Comment:
We should specify the snapshot executor for the async job

Reply:
I'm afraid not if we use the same executor somewhere inside the task. I tried. The executor might be confgured with single thred. This thread is blocked with the waiting task. No thread left for the workers. Tests like `testChangeSnapshotTransferRateInRuntime()` hangs.